batch/batch/driver/main.py

```diff
@@ -1322,6 +1330,150 @@ async def monitor_system(app):
     monitor_instances(app)
 
 
+async def compact_agg_billing_project_users_table(db: Database):
+    @transaction(db)
+    async def compact(tx: Transaction):
+        target = await tx.execute_and_fetchone(
+            '''
+SELECT billing_project, `user`, resource_id
+FROM (
+  SELECT billing_project, `user`, resource_id, COUNT(*) AS n_tokens
+  FROM aggregated_billing_project_user_resources_v3
+  WHERE token != 0
+  GROUP BY billing_project, `user`, resource_id
+) AS t
+ORDER BY n_tokens DESC
+LIMIT 1;
+''',
+            query_name='find_agg_billing_project_user_resource_to_compact',
+        )
+
+        if target is None:
+            return
+
+        original_usage = await tx.execute_and_fetchone(
+            '''
+SELECT CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s
+FOR UPDATE;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        # overwriting the value for token = 0 here is intentional because we've computed the total usage above
+        await tx.execute_update(
+            '''
+INSERT INTO aggregated_billing_project_user_resources_v3 (billing_project, `user`, resource_id, token, `usage`)
+VALUES (%s, %s, %s, %s, %s)
+ON DUPLICATE KEY UPDATE `usage` = %s
+''',
+            (target['billing_project'], target['user'], target['resource_id'], 0, original_usage, original_usage),
+        )
+
+        await tx.just_execute(
+            '''
+DELETE FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s AND token != 0;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        new_usage = await tx.execute_and_fetchone(
+            '''
+SELECT billing_project, `user`, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s
+GROUP BY billing_project, `user`, resource_id;
```

Comment:
I don't think we need this group by. Should be just like the first SELECT if I understand correctly.

Reply:
We can either specify only to get the usage for `token=0` as that **should be** the only record. However, I think this is safer in case extra records are there. At least we'll know we haven't changed the database values even if it's not quite what we wanted.

Reply:
The group by here shouldn't change the performance as the WHERE filter executes first.

Reply:
We need both selects. We need to add a FOR UPDATE to the rows we want to lock. But if we add it to the first query, then we're going to lock the entire table as it reads all rows. So the second select is really a locking point query.

Reply:
This query must have a value we already know given the semantics of MySQL, AFAICT. We first run:
```
SELECT CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
FROM aggregated_billing_project_user_resources_v3
WHERE billing_project = %s AND `user` = %s AND resource_id = %s
FOR UPDATE;
```
This computes the total usage for a given bp, user, resource id *and also* prevents any other transaction from inserting, modifying, or deleting a record with this bp, this user, and this resource id.

Then, if we make my proposed simplification above, we run this:
```
DELETE FROM aggregated_billing_project_user_resources_v3
WHERE billing_project = %s AND `user` = %s AND resource_id = %s;
```
This could only delete the records which we just summed because we're holding an exclusive lock (including any necessary gap locks) on said records.

Finally we run
```
INSERT INTO aggregated_billing_project_user_resources_v3 (billing_project, `user`, resource_id, token, `usage`)
VALUES (%s, %s, %s, %s, %s)
```
Since we know there can be no records with this bp, user, and resource id (since we have an exclusive lock and also just deleted them all), this must insert a single record without failing. We now know for an absolute fact that we have exactly one record. We know its usage number too.

At this point, this query can only return the usage we know it to be.
```
SELECT CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
FROM aggregated_billing_project_user_resources_v3
WHERE billing_project = %s AND `user` = %s AND resource_id = %s
GROUP BY billing_project, `user`, resource_id;
```